libs/dfi/src/dyn_type.c

```diff
@@ -185,265 +187,200 @@ static int dynType_parseAny(FILE *stream, dyn_type *type) {
-        entry->value = value;
-        TAILQ_INSERT_TAIL(&type->metaProperties, entry, entries);
-    } else {
-        LOG_ERROR("Failed to parse meta properties");
-        free(name);
-        free(value);
-        free(entry);
+    struct meta_entry *entry = calloc(1, sizeof(*entry));
+    if (entry == NULL) {
+        status = MEM_ERROR;
+        goto bail_out;
     }
+    entry->name = celix_steal_ptr(name);
+    entry->value = celix_steal_ptr(value);
+    TAILQ_INSERT_TAIL(&type->metaProperties, entry, entries);
+    return OK;
 
+bail_out:
+    celix_err_push("Failed to parse meta properties");
     return status;
 }
 
-static int dynType_parseText(FILE *stream, dyn_type *type) {
-    int status = OK;
+static int dynType_parseText(FILE* stream, dyn_type* type) {
     type->type = DYN_TYPE_TEXT;
     type->descriptor = 't';
+    type->trivial = false;
     type->ffiType = &ffi_type_pointer;
-    return status;
+    return OK;
 }
 
-static int dynType_parseEnum(FILE *stream, dyn_type *type) {
-    int status = OK;
+static int dynType_parseEnum(FILE* stream, dyn_type* type) {
     type->ffiType = &ffi_type_sint32;
     type->descriptor = 'E';
+    type->trivial = true;
     type->type = DYN_TYPE_SIMPLE;
-    return status;
+    return OK;
 }
 
-static int dynType_parseComplex(FILE *stream, dyn_type *type) {
+static int dynType_parseComplex(FILE* stream, dyn_type* type) {
+    size_t nbEntries = 0;
     int status = OK;
     type->type = DYN_TYPE_COMPLEX;
     type->descriptor = '{';
+    type->trivial = true;
     type->ffiType = &type->complex.structType;
     TAILQ_INIT(&type->complex.entriesHead);
 
     int c = fgetc(stream);
-    struct complex_type_entry *entry = NULL;
+    struct complex_type_entry* entry = NULL;
     while (c != ' ' && c != '}') {
         ungetc(c,stream);
+        celix_autoptr(dyn_type) subType = NULL;
+        status = dynType_parseWithStreamOfName(stream, NULL, type, NULL, &subType, dynType_parseAny);
+        if (status != OK) {
+            return status;
+        }
         entry = calloc(1, sizeof(*entry));
-        if (entry != NULL) {
-            entry->type = calloc(1, sizeof(*entry->type));
+        if (entry == NULL) {
+            celix_err_push("Error allocating memory for complex_type_entry");
+            return MEM_ERROR;
         }
-        if (entry != NULL && entry->type != NULL) {
-            entry->type->parent = type;
-            entry->type->type = DYN_TYPE_INVALID;
-            TAILQ_INIT(&entry->type->nestedTypesHead);
-            TAILQ_INIT(&entry->type->metaProperties);
-            TAILQ_INSERT_TAIL(&type->complex.entriesHead, entry, entries);
-            status = dynType_parseAny(stream, entry->type);
-        } else {
-            free(entry);
-            status = MEM_ERROR;
-            LOG_ERROR("Error allocating memory for type");
+        entry->type = celix_steal_ptr(subType);
+        if (!dynType_isTrivial(entry->type)) {
+            type->trivial = false;
```

Comment:
```suggestion
        type->trivial = dynType_isTrivial(entry->type); 
```

Reply:
The intention here is if ANY of its field is non-trivial, then the complex itself is non-trivial, otherwise it is trivial (as set in line 247).